src/cluster_legacy.c

```diff
@@ -7029,48 +7029,73 @@ int clusterCommandSpecial(client *c) {
         clusterDelNode(n);
         clusterDoBeforeSleep(CLUSTER_TODO_UPDATE_STATE | CLUSTER_TODO_SAVE_CONFIG);
         addReply(c, shared.ok);
-    } else if (!strcasecmp(c->argv[1]->ptr, "replicate") && c->argc == 3) {
+    } else if (!strcasecmp(c->argv[1]->ptr, "replicate") && (c->argc == 3 || c->argc == 4)) {
         /* CLUSTER REPLICATE <NODE ID> */
         /* Lookup the specified node in our table. */
-        clusterNode *n = clusterLookupNode(c->argv[2]->ptr, sdslen(c->argv[2]->ptr));
-        if (!n) {
-            addReplyErrorFormat(c, "Unknown node %s", (char *)c->argv[2]->ptr);
-            return 1;
+        clusterNode *n = NULL;
+        if (c->argc == 4) {
+            if (0 != strcasecmp(c->argv[2]->ptr,"NO") || 0 != strcasecmp(c->argv[3]->ptr,"ONE")) {
+                addReplySubcommandSyntaxError(c);
+                return 1;
+            }
+        } else {
+            n = clusterLookupNode(c->argv[2]->ptr, sdslen(c->argv[2]->ptr));
+            if (!n) {
+                addReplyErrorFormat(c, "Unknown node %s", (char *)c->argv[2]->ptr);
+                return 1;
+            }
         }
 
-        /* I can't replicate myself. */
-        if (n == myself) {
-            addReplyError(c, "Can't replicate myself");
-            return 1;
-        }
+        if (n == NULL) {
+            if (nodeIsPrimary(myself)) {
+                addReply(c,shared.ok);
+                return 1;
+            }
+            serverLog(LL_NOTICE,"Stop replication and turning myself into empty primary.");
+            clusterSetNodeAsPrimary(myself);
+            if (server.primary != NULL) {
+                replicationUnsetPrimary();
+            }
```

Comment:
All other invocation of `replicationUnsetPrimary()` don't have this wrapped under the condition. Is it unnecessary to invoke it if server.primary is NULL ?

Reply:
Also unsure why `clusterPromoteSelfToPrimary` was introduced, seems like it's the same behavior at this point but good to call this abstraction if there will be additional steps introduced in the future  for cluster mode.

Reply:
>Also unsure why clusterPromoteSelfToPrimary was introduced, seems like it's the same behavior at this point but good to call this abstraction if there will be additional steps introduced in the future for cluster mode.

Ok.

Reply:
> All other invocation of replicationUnsetPrimary() don't have this wrapped under the condition. Is it unnecessary to invoke it if server.primary is NULL ?

Probably not needed. It looks like left over from KeyDB's implementation where primary needed to be passed to [replicationUnsetMaster](https://github.com/Snapchat/KeyDB/blob/603ebb27fb82a27fb98b0feb6749b0f7661a1c4b/src/cluster.cpp#L4941-L4945)